debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/transforms/MongoDataConverter.java

```diff
@@ -0,0 +1,324 @@
+                            convertFieldValue(entry9, struct1, schema1);
+                        }
+                        list.add(struct1);
+                    }
+                });
+                colValue = list;
+            }
+            break;
+
+        default:
+            break;
+        }
+        struct.put(key, keyvalueforStruct.getValue().isNull() ? null : colValue);
+    }
+
+    public static void addFieldSchema(Entry<String, BsonValue> keyValuesforSchema, SchemaBuilder builder) {
+        String key = keyValuesforSchema.getKey();
+        BsonType type = keyValuesforSchema.getValue().getBsonType();
+
+        switch (type) {
+
+        case NULL:
+            LOG.warn("Data type {} not currently supported", type);
+            break;
+
+        case STRING:
+        case JAVASCRIPT:
+        case OBJECT_ID:
+        case DECIMAL128:
+            builder.field(key, Schema.OPTIONAL_STRING_SCHEMA);
+            break;
+
+        case DOUBLE:
+            builder.field(key, Schema.OPTIONAL_FLOAT64_SCHEMA);
+            break;
+
+        case BINARY:
+            builder.field(key, Schema.OPTIONAL_BYTES_SCHEMA);
+            break;
+
+        case INT32:
+        case TIMESTAMP:
+            builder.field(key, Schema.OPTIONAL_INT32_SCHEMA);
+            break;
+
+        case INT64:
+        case DATE_TIME:
+            builder.field(key, Schema.OPTIONAL_INT64_SCHEMA);
+            break;
+
+        case BOOLEAN:
+            builder.field(key, Schema.OPTIONAL_BOOLEAN_SCHEMA);
+            break;
+
+        case JAVASCRIPT_WITH_SCOPE:
+            SchemaBuilder jswithscope = SchemaBuilder.struct();
+            jswithscope.field("code", Schema.OPTIONAL_STRING_SCHEMA);
+            SchemaBuilder scope = SchemaBuilder.struct();
+            BsonDocument jwsDocument = keyValuesforSchema.getValue().asJavaScriptWithScope().getScope().asDocument();
+
+            for (Entry<String, BsonValue> jwsDocumentKey : jwsDocument.entrySet()) {
+                addFieldSchema(jwsDocumentKey, scope);
+            }
+
+            Schema scopeBuild = scope.build();
+            jswithscope.field("scope", scopeBuild).build();
+            builder.field(key, jswithscope);
+            break;
+
+        case REGULAR_EXPRESSION:
+            SchemaBuilder regexwop = SchemaBuilder.struct();
+            regexwop.field("regex", Schema.OPTIONAL_STRING_SCHEMA);
+            regexwop.field("options", Schema.OPTIONAL_STRING_SCHEMA);
+            builder.field(key, regexwop.build());
+            break;
+
+        case DOCUMENT:
+            SchemaBuilder builderDoc = SchemaBuilder.struct();
+            BsonDocument docs = keyValuesforSchema.getValue().asDocument();
+
+            for (Entry<String, BsonValue> doc : docs.entrySet()) {
+                addFieldSchema(doc, builderDoc);
+            }
+            builder.field(key, builderDoc.build());
+            break;
+
+        case ARRAY:
```

Comment:
Will this work with a deeply nested structure? I.e. an array of arrays? I think this ought to contain some kind of recursiveness really.

Reply:
I've filed https://issues.jboss.org/browse/DBZ-562 as a follow-up for that one.